File: LeetCode/Medium/0973-k-closest-points-to-origin/0973-k-closest-points-to-origin.java
class Solution {
    public int[][] kClosest(int[][] points, int k) {
        PriorityQueue<int[]> pq = new PriorityQueue<>((o1, o2) -> o2[0] - o1[0]);

        for (int[] p : points) {
            int d = p[0] * p[0] + p[1] * p[1];
            pq.offer(new int[] {d, p[0], p[1]});

            if (pq.size() > k) {
                pq.poll();
            }
        }

        int[][] arr = new int[k][];

        while (k > 0) {
            int[] v = pq.poll();
            arr[--k] = new int[] {v[1], v[2]};
        }
        
        return arr;
    }
}

File: LeetCode/Medium/0973-k-closest-points-to-origin/0973-k-closest-points-to-origin.kt
class Solution {
    fun kClosest(points: Array<IntArray>, k: Int): Array<IntArray> {
        val pq = PriorityQueue<IntArray>(compareByDescending { it[0]} )

        for (p in points) {
            val d = p[0] * p[0] + p[1] * p[1]
            pq.offer(intArrayOf(d, p[0], p[1]))

            if (pq.size > k) {
                pq.poll()
            }
        }

        val arr = Array(k) { IntArray(2) }
        
        for (i in 0 until k) {
            val v = pq.poll()
            arr[i] = intArrayOf(v[1], v[2])
        }

        return arr
    }
}